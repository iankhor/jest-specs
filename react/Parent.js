import React, { Component } from 'react'
import request from './../async/request'

const Child = props => (
  <div>
    <input id="child-click" onClick={props.childOnClick}>Child 1</input>
    <input id="child-async-click" onClick={props.requestOnClick}>Child 2</input>
    <GrandChild grandChildOnClick={props.grandChildOnClick}/>
  </div>
)

const GrandChild = props => (
  <div>
    <input id="grand-child-click" onClick={() => props.grandChildOnClick(3)}>grandChild</input>
  </div>
)

export { Child, GrandChild }

export default class Parent extends Component {
  constructor(props) {
    super(props)

    this.state = {
      childValue: false,
      grandChildValue: 0,
      asyncResponse: ''
    }
  }

  request = async () => {
    const response = await request('mojo')

    this.setState({ asyncResponse: response })
  }

  childOnClick = () => this.setState({ childValue: true })

  grandChildOnClick = value => {
    this.setState({ grandChildValue: value })
  }

  render() {
    return(
      <div>
        <input id="click" onClick={this.props.someFunction}></input>
        <Child
          childOnClick={this.childOnClick}
          requestOnClick={this.request}
          grandChildOnClick={this.grandChildOnClick}
        />
      </div>
    )
  }
}
