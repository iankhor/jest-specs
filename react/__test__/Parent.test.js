import React from 'react'
import Parent from './../Parent'
import { shallow } from 'enzyme'

describe('Parent', () => {
  let component
  let someFunction = jest.fn()

  beforeEach(() => component = shallow(<Parent someFunction={someFunction}/>))

  it('renders', () => expect(component).toHaveLength(1))

  describe('someFunction', () => {
    describe('when input element is clicked once', () => {
      afterEach(() => someFunction.mockReset())

      it('called once', () => {
        component.find('input[id="click"]').simulate('click')

        expect(someFunction).toHaveBeenCalledTimes(1);
      })
    })

    describe('when input element is clicked twice', () => {
      it('called once', () => {
        component.find('input[id="click"]').simulate('click')
        component.find('input[id="click"]').simulate('click')

        expect(someFunction).toHaveBeenCalledTimes(2);
      })
    })
  })

  describe('Child', () => {
    let childComponent

    beforeEach(() => childComponent = component.find('Child').dive())

    it('renders', () => expect(component.find('Child')).toHaveLength(1))

    it('updates childValue if clicked on', () => {
      childComponent.find('input[id="child-click"]').simulate('click')

      expect(component.state().childValue).toEqual(true)
    })

    it('updates asyncResponse if clicked on', (done) => {
      childComponent.find('input[id="child-async-click"]').simulate('click')

      process.nextTick(() => {
        expect(component.state().asyncResponse).toEqual('real request: mojo')
        done()
      })
    })
  })
})
