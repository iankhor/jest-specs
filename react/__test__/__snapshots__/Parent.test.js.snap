// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <input
    id="child-click"
  >
    Child 1
  </input>
  <input
    id="child-async-click"
  >
    Child 2
  </input>
  <div>
    <input
      id="grand-child-click"
      onClick={[Function]}
    >
      grandChild
    </input>
  </div>
</div>
`;
